libs/dagrun/dagrun.go

```diff
@@ -170,34 +175,71 @@ func (g *Graph[N]) Run(pool int, runUnit func(N)) {
 		panic("dagrun: no entry points")
 	}
 
-	ready := make(chan N, len(in))
-	done := make(chan N, len(in))
+	// For each node, remember a failed direct dependency (any one) if present.
+	failedCause := make(map[N]*N, len(in))
+
+	ready := make(chan task[N], len(in))
+	done := make(chan doneResult[N], len(in))
 
 	var wg sync.WaitGroup
 	wg.Add(pool)
 	for range pool {
-		go func() {
-			defer wg.Done()
-			for n := range ready {
-				runUnit(n)
-				done <- n
-			}
-		}()
+		go runWorkerLoop(&wg, ready, done, runUnit)
 	}
 
 	for _, n := range initial {
-		ready <- n
+		ready <- task[N]{n: n, failedFrom: nil}
 	}
 
-	for remaining := len(in); remaining > 0; {
-		n := <-done
-		remaining--
-		for _, e := range g.adj[n] {
+	for remaining := len(in); remaining > 0; remaining-- {
+		res := <-done
+
+		if !res.success {
+			// Determine the originating failure cause to propagate
+			var cause *N
+			if existing, ok := failedCause[res.n]; ok && existing != nil {
+				cause = existing
```

Comment:
It does not need to happen now, but we could extend it to return all failed deps in the future.

Reply:
did not get it -- what for?